substrate/in-instructions/primitives/src/shorthand.rs

```diff
@@ -0,0 +1,42 @@
+use scale::{Encode, Decode, MaxEncodedLen};
+use scale_info::TypeInfo;
+
+#[cfg(feature = "std")]
+use serde::{Serialize, Deserialize};
+
+use sp_core::{ConstU32, bounded::BoundedVec};
+
+use serai_primitives::{NativeAddress, Coin, Amount};
+
+use crate::{MAX_DATA_LEN, ExternalAddress, ExternalInInstruction, OutInstruction};
+
+#[derive(Clone, PartialEq, Eq, Debug, Encode, Decode, MaxEncodedLen, TypeInfo)]
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+pub enum Shorthand {
+  Raw(BoundedVec<u8, ConstU32<{ MAX_DATA_LEN }>>),
+  Swap {
+    origin: Option<ExternalAddress>,
+    coin: Coin,
+    minimum: Amount,
+    out: OutInstruction,
+  },
+  AddLiquidity {
+    origin: Option<ExternalAddress>,
+    minimum: Amount,
```

Comment:
no minimum needed for add liq.

Reply:
Minimum LP tokens. It prevents adding at expected ratio X:Y, having it frontrun to 2X:Y, yet getting credited for X:0.5y which your amount now is.

Reply:
Ok but if we are not doing single sided add there's no swap interaction and the user must have/send TXs for both.

Reply:
Right, yet that'd be a mint from BTC/XMR, not a Shorthand AddLiquidity, then a direct call on Serai.

Reply:
This does not make sense. You want liquidity to be asym deposit only?

Reply:
No. The flow for adding sym liq should be as follows:

- Mint sriX from X
- From an account with sriX and SRI, add liquidity.

This instruction is for the path where you don't have SRI and only want to add liquidity for X from X.